pwiz_tools/Shared/Common/Chemistry/MoleculeMassOffset.cs

```diff
@@ -29,73 +29,197 @@ namespace pwiz.Common.Chemistry
-                                         ToSignedString(format, formatProvider, AverageMassOffset) + ")");
-                }
+                monoMassOffset = TypedMass.ZERO_MONO_MASSNEUTRAL;
+            }
+            if (averageMassOffset == null)
+            {
+                averageMassOffset = TypedMass.ZERO_AVERAGE_MASSNEUTRAL;
             }
 
-            return stringBuilder.ToString();
+            return Molecule.IsNullOrEmpty(molecule) &&
+                   monoMassOffset== 0 && averageMassOffset == 0 ?
+                EMPTY :
+                new MoleculeMassOffset(molecule, monoMassOffset, averageMassOffset);
+        }
+
+        // N.B. The mass portion is always written with InvariantCulture.
+
+        public static string FormatMassModification(double massMod, int desiredDecimals)
+        {
+            if (massMod == 0)
+            {
+                return string.Empty;
+            }
+            var sign = massMod > 0 ? @"+" : string.Empty;
+            return string.Format($@"[{sign}{massMod.ToString($"F{desiredDecimals}", CultureInfo.InvariantCulture).TrimEnd('0')}]");
         }
 
-        private static string ToSignedString(string format, IFormatProvider formatProvider, double value)
+        public static string FormatMassModification(double massModMono, double massModAverage, int desiredDecimals)
         {
-            return (value >= 0 ? @"+" : string.Empty) + value.ToString(format, formatProvider);
+            if (massModMono == 0 && massModAverage == 0)
+            {
+                return string.Empty;
+            }
+            if (Equals(massModMono, massModAverage))
+            {
+                return FormatMassModification(massModMono, desiredDecimals);
+            }
+            var sign = massModMono > 0 ? @"+" : string.Empty;
+            return string.Format($@"[{sign}{massModMono.ToString($"F{desiredDecimals}", CultureInfo.InvariantCulture).TrimEnd('0')}/{Math.Abs(massModAverage).ToString($"F{desiredDecimals}", CultureInfo.InvariantCulture).TrimEnd('0')}]");
+        }
+
+        public override string ToString()
+        {
+            return ToString(null, CultureInfo.CurrentCulture);
+        }
+
+        // Satisfy IFormattable interface
+        public string ToString(string format, IFormatProvider formatProvider)
+        {
+            return ToString(6);
+        }
+
+        public virtual string ChemicalFormulaString() => Molecule.ToString();
+        
+        public string ToString(int desiredDigits)
+        {
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append(ChemicalFormulaString());
+            stringBuilder.Append(FormatMassModification(MonoMassOffset, AverageMassOffset, desiredDigits)); // Use our standard notation e.g. [+1.23/1.24], [-1.2] etc
+            return stringBuilder.ToString();
         }
 
-        protected bool Equals(MoleculeMassOffset other)
+        public bool Equals(MoleculeMassOffset other)
```

Comment:
Hmm, how do you satisfy IEquitable<> then?

Reply:
Good question.

This [web page](https://www.c-sharpcorner.com/article/story-of-equality-in-net-part-three/) seems to imply that the `IEquatable<T>` interface is probably only useful for "struct" types. There is some overhead in converting a struct to an object, so for struct types, it is faster to call the `IEquatable<T>.Equals` method if it exists instead of the regular `Equals(Object)` method.
This apparently makes `Dictionary<TKey, TValue>` faster. Dictionary uses `System.Collections.Generic.EqualityComparer<T>.Default` which checks whether "T" implements `IEquatable<T>`.

I'm thinking we should only be implementing IEquatable&lt;T> on our struct's, and if we do implement it on a class we need to either make it virtual or make sure it calls a virtual method because otherwise it will be ignoring fields declared in subclasses.